src/rgw/s3select_csv_parser.h

```diff
@@ -0,0 +1,332 @@
+#include <iostream>
+
+#include <boost/mpl/vector/vector30.hpp>
+// back-end
+#include <boost/msm/back/state_machine.hpp>
+//front-end
+#include <boost/msm/front/state_machine_def.hpp>
+
+#include <vector>
+
+namespace msm = boost::msm;
+namespace mpl = boost::mpl;
+
+namespace s3selectEngine
+{
+	// events
+	struct event_column_sep {};
+	struct event_eol {};
+	struct event_end_of_stream {};
+	struct event_not_column_sep {};//i.e any char
+	struct event_quote {};
+	struct event_escape {};
+	struct event_empty {};
+
+
+	// front-end: define the FSM structure 
+	struct csvStateMch_ : public msm::front::state_machine_def<csvStateMch_>
+	{
+		char * input_stream;
+		std::vector<char*> *tokens;
+		std::vector<int> has_esc{128};
+		size_t token_idx;
+		size_t escape_idx;
+		char *input_cur_location;
+		char *start_token;
+		bool end_of_parse;
+
+		typedef csvStateMch_ csv_rules;
+
+		csvStateMch_():end_of_parse(false){}
+
+		void set(const char * input, std::vector<char*> *tk)
+		{
+			input_cur_location = input_stream = (char*)input;
+			token_idx = 0;
+			tokens = tk;
+			escape_idx = 0;
+		}
+
+		char get_char(){return *input_cur_location;}
+
+		char get_next_char(){input_cur_location++;return *input_cur_location;}
+
+		const char* currentLoc(){return input_cur_location;}
+
+		void parse_escape(char *in,char esc_char='\\')
+		{
+			//assumption atleast one escape and single
+			char *dst, *src;
+
+			dst = src = in;
+
+			while (1)
+			{
+				while (*src && *src != esc_char)
```

Comment:
not very c++ like, but

Reply:
indeed it's c, but perform well.
in the case of CSV input contain escape-rules , it means "shrinking" the input csv  (kind of copy and paste "into itself")
escape-char + char  BECOME one-char
it impact parsing performance 

this "helper" routine , is for the second-pass , upon parsing is ended, and tokens containing escape-rule are identified and processed again (escape conversion)

letting the state machine do the all process (upon identified escape rule , do copy and paste), might impact parser performance